File: clients/java/src/main/java/ai_cup_22/model/WeaponProperties.java
package ai_cup_22.model;

import ai_cup_22.util.StreamUtil;

/**
 * Weapon properties
 */
public class WeaponProperties {
    /**
     * Name
     */
    private String name;

    /**
     * Name
     */
    public String getName() {
        return name;
    }

    /**
     * Name
     */
    public void setName(String value) {
        this.name = value;
    }
    /**
     * Shooting speed (number of shots per second)
     */
    private double roundsPerSecond;

    /**
     * Shooting speed (number of shots per second)
     */
    public double getRoundsPerSecond() {
        return roundsPerSecond;
    }

    /**
     * Shooting speed (number of shots per second)
     */
    public void setRoundsPerSecond(double value) {
        this.roundsPerSecond = value;
    }
    /**
     * Accuracy (spread angle) of a shot (in degrees)
     */
    private double spread;

    /**
     * Accuracy (spread angle) of a shot (in degrees)
     */
    public double getSpread() {
        return spread;
    }

    /**
     * Accuracy (spread angle) of a shot (in degrees)
     */
    public void setSpread(double value) {
        this.spread = value;
    }
    /**
     * Aiming time
     */
    private double aimTime;

    /**
     * Aiming time
     */
    public double getAimTime() {
        return aimTime;
    }

    /**
     * Aiming time
     */
    public void setAimTime(double value) {
        this.aimTime = value;
    }
    /**
     * Field of view in full aim (in degrees)
     */
    private double aimFieldOfView;

    /**
     * Field of view in full aim (in degrees)
     */
    public double getAimFieldOfView() {
        return aimFieldOfView;
    }

    /**
     * Field of view in full aim (in degrees)
     */
    public void setAimFieldOfView(double value) {
        this.aimFieldOfView = value;
    }
    /**
     * Rotation speed in full aim (degrees per second)
     */
    private double aimRotationSpeed;

    /**
     * Rotation speed in full aim (degrees per second)
     */
    public double getAimRotationSpeed() {
        return aimRotationSpeed;
    }

    /**
     * Rotation speed in full aim (degrees per second)
     */
    public void setAimRotationSpeed(double value) {
        this.aimRotationSpeed = value;
    }
    /**
     * Movement speed modifier in full aim
     */
    private double aimMovementSpeedModifier;

    /**
     * Movement speed modifier in full aim
     */
    public double getAimMovementSpeedModifier() {
        return aimMovementSpeedModifier;
    }

    /**
     * Movement speed modifier in full aim
     */
    public void setAimMovementSpeedModifier(double value) {
        this.aimMovementSpeedModifier = value;
    }
    /**
     * Speed of projectiles
     */
    private double projectileSpeed;

    /**
     * Speed of projectiles
     */
    public double getProjectileSpeed() {
        return projectileSpeed;
    }

    /**
     * Speed of projectiles
     */
    public void setProjectileSpeed(double value) {
        this.projectileSpeed = value;
    }
    /**
     * Damage of a projectile
     */
    private double projectileDamage;

    /**
     * Damage of a projectile
     */
    public double getProjectileDamage() {
        return projectileDamage;
    }

    /**
     * Damage of a projectile
     */
    public void setProjectileDamage(double value) {
        this.projectileDamage = value;
    }
    /**
     * Projectiles' life time
     */
    private double projectileLifeTime;

    /**
     * Projectiles' life time
     */
    public double getProjectileLifeTime() {
        return projectileLifeTime;
    }

    /**
     * Projectiles' life time
     */
    public void setProjectileLifeTime(double value) {
        this.projectileLifeTime = value;
    }
    /**
     * Index of the sound when shooting (starting with 0), or None
     */
    private Integer shotSoundTypeIndex;

    /**
     * Index of the sound when shooting (starting with 0), or None
     */
    public Integer getShotSoundTypeIndex() {
        return shotSoundTypeIndex;
    }

    /**
     * Index of the sound when shooting (starting with 0), or None
     */
    public void setShotSoundTypeIndex(Integer value) {
        this.shotSoundTypeIndex = value;
    }
    /**
     * Index of the sound when hitting something (starting with 0), or None
     */
    private Integer projectileHitSoundTypeIndex;

    /**
     * Index of the sound when hitting something (starting with 0), or None
     */
    public Integer getProjectileHitSoundTypeIndex() {
        return projectileHitSoundTypeIndex;
    }

    /**
     * Index of the sound when hitting something (starting with 0), or None
     */
    public void setProjectileHitSoundTypeIndex(Integer value) {
        this.projectileHitSoundTypeIndex = value;
    }
    /**
     * Max amount of ammo unit can hold in their inventory
     */
    private int maxInventoryAmmo;

    /**
     * Max amount of ammo unit can hold in their inventory
     */
    public int getMaxInventoryAmmo() {
        return maxInventoryAmmo;
    }

    /**
     * Max amount of ammo unit can hold in their inventory
     */
    public void setMaxInventoryAmmo(int value) {
        this.maxInventoryAmmo = value;
    }

    public WeaponProperties(String name, double roundsPerSecond, double spread, double aimTime, double aimFieldOfView, double aimRotationSpeed, double aimMovementSpeedModifier, double projectileSpeed, double projectileDamage, double projectileLifeTime, Integer shotSoundTypeIndex, Integer projectileHitSoundTypeIndex, int maxInventoryAmmo) {
        this.name = name;
        this.roundsPerSecond = roundsPerSecond;
        this.spread = spread;
        this.aimTime = aimTime;
        this.aimFieldOfView = aimFieldOfView;
        this.aimRotationSpeed = aimRotationSpeed;
        this.aimMovementSpeedModifier = aimMovementSpeedModifier;
        this.projectileSpeed = projectileSpeed;
        this.projectileDamage = projectileDamage;
        this.projectileLifeTime = projectileLifeTime;
        this.shotSoundTypeIndex = shotSoundTypeIndex;
        this.projectileHitSoundTypeIndex = projectileHitSoundTypeIndex;
        this.maxInventoryAmmo = maxInventoryAmmo;
    }

    /**
     * Read WeaponProperties from input stream
     */
    public static WeaponProperties readFrom(java.io.InputStream stream) throws java.io.IOException {
        String name;
        name = StreamUtil.readString(stream);
        double roundsPerSecond;
        roundsPerSecond = StreamUtil.readDouble(stream);
        double spread;
        spread = StreamUtil.readDouble(stream);
        double aimTime;
        aimTime = StreamUtil.readDouble(stream);
        double aimFieldOfView;
        aimFieldOfView = StreamUtil.readDouble(stream);
        double aimRotationSpeed;
        aimRotationSpeed = StreamUtil.readDouble(stream);
        double aimMovementSpeedModifier;
        aimMovementSpeedModifier = StreamUtil.readDouble(stream);
        double projectileSpeed;
        projectileSpeed = StreamUtil.readDouble(stream);
        double projectileDamage;
        projectileDamage = StreamUtil.readDouble(stream);
        double projectileLifeTime;
        projectileLifeTime = StreamUtil.readDouble(stream);
        Integer shotSoundTypeIndex;
        if (StreamUtil.readBoolean(stream)) {
            shotSoundTypeIndex = StreamUtil.readInt(stream);
        } else {
            shotSoundTypeIndex = null;
        }
        Integer projectileHitSoundTypeIndex;
        if (StreamUtil.readBoolean(stream)) {
            projectileHitSoundTypeIndex = StreamUtil.readInt(stream);
        } else {
            projectileHitSoundTypeIndex = null;
        }
        int maxInventoryAmmo;
        maxInventoryAmmo = StreamUtil.readInt(stream);
        return new WeaponProperties(name, roundsPerSecond, spread, aimTime, aimFieldOfView, aimRotationSpeed, aimMovementSpeedModifier, projectileSpeed, projectileDamage, projectileLifeTime, shotSoundTypeIndex, projectileHitSoundTypeIndex, maxInventoryAmmo);
    }

    /**
     * Write WeaponProperties to output stream
     */
    public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
        StreamUtil.writeString(stream, name);
        StreamUtil.writeDouble(stream, roundsPerSecond);
        StreamUtil.writeDouble(stream, spread);
        StreamUtil.writeDouble(stream, aimTime);
        StreamUtil.writeDouble(stream, aimFieldOfView);
        StreamUtil.writeDouble(stream, aimRotationSpeed);
        StreamUtil.writeDouble(stream, aimMovementSpeedModifier);
        StreamUtil.writeDouble(stream, projectileSpeed);
        StreamUtil.writeDouble(stream, projectileDamage);
        StreamUtil.writeDouble(stream, projectileLifeTime);
        if (shotSoundTypeIndex == null) {
            StreamUtil.writeBoolean(stream, false);
        } else {
            StreamUtil.writeBoolean(stream, true);
            StreamUtil.writeInt(stream, shotSoundTypeIndex);
        }
        if (projectileHitSoundTypeIndex == null) {
            StreamUtil.writeBoolean(stream, false);
        } else {
            StreamUtil.writeBoolean(stream, true);
            StreamUtil.writeInt(stream, projectileHitSoundTypeIndex);
        }
        StreamUtil.writeInt(stream, maxInventoryAmmo);
    }

    /**
     * Get string representation of WeaponProperties
     */
    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder("WeaponProperties { ");
        stringBuilder.append("name: ");
        stringBuilder.append('"' + name + '"');
        stringBuilder.append(", ");
        stringBuilder.append("roundsPerSecond: ");
        stringBuilder.append(String.valueOf(roundsPerSecond));
        stringBuilder.append(", ");
        stringBuilder.append("spread: ");
        stringBuilder.append(String.valueOf(spread));
        stringBuilder.append(", ");
        stringBuilder.append("aimTime: ");
        stringBuilder.append(String.valueOf(aimTime));
        stringBuilder.append(", ");
        stringBuilder.append("aimFieldOfView: ");
        stringBuilder.append(String.valueOf(aimFieldOfView));
        stringBuilder.append(", ");
        stringBuilder.append("aimRotationSpeed: ");
        stringBuilder.append(String.valueOf(aimRotationSpeed));
        stringBuilder.append(", ");
        stringBuilder.append("aimMovementSpeedModifier: ");
        stringBuilder.append(String.valueOf(aimMovementSpeedModifier));
        stringBuilder.append(", ");
        stringBuilder.append("projectileSpeed: ");
        stringBuilder.append(String.valueOf(projectileSpeed));
        stringBuilder.append(", ");
        stringBuilder.append("projectileDamage: ");
        stringBuilder.append(String.valueOf(projectileDamage));
        stringBuilder.append(", ");
        stringBuilder.append("projectileLifeTime: ");
        stringBuilder.append(String.valueOf(projectileLifeTime));
        stringBuilder.append(", ");
        stringBuilder.append("shotSoundTypeIndex: ");
        stringBuilder.append(String.valueOf(shotSoundTypeIndex));
        stringBuilder.append(", ");
        stringBuilder.append("projectileHitSoundTypeIndex: ");
        stringBuilder.append(String.valueOf(projectileHitSoundTypeIndex));
        stringBuilder.append(", ");
        stringBuilder.append("maxInventoryAmmo: ");
        stringBuilder.append(String.valueOf(maxInventoryAmmo));
        stringBuilder.append(" }");
        return stringBuilder.toString();
    }
}

File: clients/kotlin/src/main/kotlin/ai_cup_22/model/WeaponProperties.kt
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Weapon properties
 */
class WeaponProperties {
    /**
     * Name
     */
    var name: String
    /**
     * Shooting speed (number of shots per second)
     */
    var roundsPerSecond: Double
    /**
     * Accuracy (spread angle) of a shot (in degrees)
     */
    var spread: Double
    /**
     * Aiming time
     */
    var aimTime: Double
    /**
     * Field of view in full aim (in degrees)
     */
    var aimFieldOfView: Double
    /**
     * Rotation speed in full aim (degrees per second)
     */
    var aimRotationSpeed: Double
    /**
     * Movement speed modifier in full aim
     */
    var aimMovementSpeedModifier: Double
    /**
     * Speed of projectiles
     */
    var projectileSpeed: Double
    /**
     * Damage of a projectile
     */
    var projectileDamage: Double
    /**
     * Projectiles' life time
     */
    var projectileLifeTime: Double
    /**
     * Index of the sound when shooting (starting with 0), or None
     */
    var shotSoundTypeIndex: Int?
    /**
     * Index of the sound when hitting something (starting with 0), or None
     */
    var projectileHitSoundTypeIndex: Int?
    /**
     * Max amount of ammo unit can hold in their inventory
     */
    var maxInventoryAmmo: Int

    constructor(name: String, roundsPerSecond: Double, spread: Double, aimTime: Double, aimFieldOfView: Double, aimRotationSpeed: Double, aimMovementSpeedModifier: Double, projectileSpeed: Double, projectileDamage: Double, projectileLifeTime: Double, shotSoundTypeIndex: Int?, projectileHitSoundTypeIndex: Int?, maxInventoryAmmo: Int) {
        this.name = name
        this.roundsPerSecond = roundsPerSecond
        this.spread = spread
        this.aimTime = aimTime
        this.aimFieldOfView = aimFieldOfView
        this.aimRotationSpeed = aimRotationSpeed
        this.aimMovementSpeedModifier = aimMovementSpeedModifier
        this.projectileSpeed = projectileSpeed
        this.projectileDamage = projectileDamage
        this.projectileLifeTime = projectileLifeTime
        this.shotSoundTypeIndex = shotSoundTypeIndex
        this.projectileHitSoundTypeIndex = projectileHitSoundTypeIndex
        this.maxInventoryAmmo = maxInventoryAmmo
    }

    /**
     * Write WeaponProperties to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeString(stream, name)
        StreamUtil.writeDouble(stream, roundsPerSecond)
        StreamUtil.writeDouble(stream, spread)
        StreamUtil.writeDouble(stream, aimTime)
        StreamUtil.writeDouble(stream, aimFieldOfView)
        StreamUtil.writeDouble(stream, aimRotationSpeed)
        StreamUtil.writeDouble(stream, aimMovementSpeedModifier)
        StreamUtil.writeDouble(stream, projectileSpeed)
        StreamUtil.writeDouble(stream, projectileDamage)
        StreamUtil.writeDouble(stream, projectileLifeTime)
        val shotSoundTypeIndexValue = shotSoundTypeIndex
        if (shotSoundTypeIndexValue == null) {
            StreamUtil.writeBoolean(stream, false)
        } else {
            StreamUtil.writeBoolean(stream, true)
            StreamUtil.writeInt(stream, shotSoundTypeIndexValue)
        }
        val projectileHitSoundTypeIndexValue = projectileHitSoundTypeIndex
        if (projectileHitSoundTypeIndexValue == null) {
            StreamUtil.writeBoolean(stream, false)
        } else {
            StreamUtil.writeBoolean(stream, true)
            StreamUtil.writeInt(stream, projectileHitSoundTypeIndexValue)
        }
        StreamUtil.writeInt(stream, maxInventoryAmmo)
    }

    /**
     * Get string representation of WeaponProperties
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("WeaponProperties { ")
        stringBuilder.append("name: ")
        stringBuilder.append('"' + name + '"')
        stringBuilder.append(", ")
        stringBuilder.append("roundsPerSecond: ")
        stringBuilder.append(roundsPerSecond)
        stringBuilder.append(", ")
        stringBuilder.append("spread: ")
        stringBuilder.append(spread)
        stringBuilder.append(", ")
        stringBuilder.append("aimTime: ")
        stringBuilder.append(aimTime)
        stringBuilder.append(", ")
        stringBuilder.append("aimFieldOfView: ")
        stringBuilder.append(aimFieldOfView)
        stringBuilder.append(", ")
        stringBuilder.append("aimRotationSpeed: ")
        stringBuilder.append(aimRotationSpeed)
        stringBuilder.append(", ")
        stringBuilder.append("aimMovementSpeedModifier: ")
        stringBuilder.append(aimMovementSpeedModifier)
        stringBuilder.append(", ")
        stringBuilder.append("projectileSpeed: ")
        stringBuilder.append(projectileSpeed)
        stringBuilder.append(", ")
        stringBuilder.append("projectileDamage: ")
        stringBuilder.append(projectileDamage)
        stringBuilder.append(", ")
        stringBuilder.append("projectileLifeTime: ")
        stringBuilder.append(projectileLifeTime)
        stringBuilder.append(", ")
        stringBuilder.append("shotSoundTypeIndex: ")
        stringBuilder.append(shotSoundTypeIndex)
        stringBuilder.append(", ")
        stringBuilder.append("projectileHitSoundTypeIndex: ")
        stringBuilder.append(projectileHitSoundTypeIndex)
        stringBuilder.append(", ")
        stringBuilder.append("maxInventoryAmmo: ")
        stringBuilder.append(maxInventoryAmmo)
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read WeaponProperties from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): WeaponProperties {
            var name: String
            name = StreamUtil.readString(stream)
            var roundsPerSecond: Double
            roundsPerSecond = StreamUtil.readDouble(stream)
            var spread: Double
            spread = StreamUtil.readDouble(stream)
            var aimTime: Double
            aimTime = StreamUtil.readDouble(stream)
            var aimFieldOfView: Double
            aimFieldOfView = StreamUtil.readDouble(stream)
            var aimRotationSpeed: Double
            aimRotationSpeed = StreamUtil.readDouble(stream)
            var aimMovementSpeedModifier: Double
            aimMovementSpeedModifier = StreamUtil.readDouble(stream)
            var projectileSpeed: Double
            projectileSpeed = StreamUtil.readDouble(stream)
            var projectileDamage: Double
            projectileDamage = StreamUtil.readDouble(stream)
            var projectileLifeTime: Double
            projectileLifeTime = StreamUtil.readDouble(stream)
            var shotSoundTypeIndex: Int?
            if (StreamUtil.readBoolean(stream)) {
                shotSoundTypeIndex = StreamUtil.readInt(stream)
            } else {
                shotSoundTypeIndex = null
            }
            var projectileHitSoundTypeIndex: Int?
            if (StreamUtil.readBoolean(stream)) {
                projectileHitSoundTypeIndex = StreamUtil.readInt(stream)
            } else {
                projectileHitSoundTypeIndex = null
            }
            var maxInventoryAmmo: Int
            maxInventoryAmmo = StreamUtil.readInt(stream)
            return WeaponProperties(name, roundsPerSecond, spread, aimTime, aimFieldOfView, aimRotationSpeed, aimMovementSpeedModifier, projectileSpeed, projectileDamage, projectileLifeTime, shotSoundTypeIndex, projectileHitSoundTypeIndex, maxInventoryAmmo)
        }
    }
}